kasa/discover.py

```diff
@@ -350,99 +362,123 @@ def _get_device_class(info: dict) -> Type[SmartDevice]:
         try:
             info = json_loads(TPLinkSmartHomeProtocol.decrypt(data))
         except Exception as ex:
             raise SmartDeviceException(
-                f"Unable to read response from device: {ip}: {ex}"
+                f"Unable to read response from device: {config.host}: {ex}"
             ) from ex
 
-        _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
+        _LOGGER.debug("[DISCOVERY] %s << %s", config.host, info)
 
         device_class = Discover._get_device_class(info)
-        device = device_class(ip, port=port)
+        device = device_class(config.host, config=config)
+        sys_info = info["system"]["get_sysinfo"]
+        if (device_type := sys_info.get("mic_type")) or (
+            device_type := sys_info.get("type")
+        ):
+            config.connection_type = ConnectionType.from_values(
+                device_family=device_type, encryption_type=EncryptType.Xor.value
+            )
+        device.protocol = get_protocol(config)  # type: ignore[assignment]
         device.update_from_discover_info(info)
         return device
 
     @staticmethod
     def _get_device_instance(
-        data: bytes, ip: str, port: int, credentials: Credentials
+        data: bytes,
+        config: DeviceConfig,
     ) -> SmartDevice:
         """Get SmartDevice from the new 20002 response."""
         try:
             info = json_loads(data[16:])
-            discovery_result = DiscoveryResult(**info["result"])
         except Exception as ex:
+            _LOGGER.debug("Got invalid response from device %s: %s", config.host, data)
+            raise SmartDeviceException(
+                f"Unable to read response from device: {config.host}: {ex}"
+            ) from ex
+        try:
+            discovery_result = DiscoveryResult(**info["result"])
+        except ValidationError as ex:
+            _LOGGER.debug(
+                "Unable to parse discovery from device %s: %s", config.host, info
+            )
             raise UnsupportedDeviceException(
-                f"Unable to read response from device: {ip}: {ex}"
+                f"Unable to parse discovery from device: {config.host}: {ex}"
             ) from ex
 
         type_ = discovery_result.device_type
-        encrypt_type_ = (
-            f"{type_.split('.')[0]}.{discovery_result.mgt_encrypt_schm.encrypt_type}"
-        )
 
-        if (device_class := get_device_class_from_type_name(type_)) is None:
+        try:
+            config.connection_type = ConnectionType.from_values(
+                type_, discovery_result.mgt_encrypt_schm.encrypt_type
+            )
+        except SmartDeviceException as ex:
+            raise UnsupportedDeviceException(
+                f"Unsupported device {config.host} of type {type_} "
+                + f"with encrypt_type {discovery_result.mgt_encrypt_schm.encrypt_type}",
+                discovery_result=discovery_result.get_dict(),
+            ) from ex
```

Comment:
I'm wondering if the `from_values()` should already raise the `UnsupportedDeviceException` to avoid this re-raise? That would require some rewiring though, which may not be wanted, so just a suggestion.

Reply:
There could be a fair bit of rewiring and when the same function is called from outside of discovery it may not always be an error due to Unsupported, i.e. it could be due to invalid serialization etc.